File: react-app/src/interfaces/Application.ts
import { SexEnum } from "../enums/SexEnum";
import { StatusEnum } from "../enums/StatusEnum";

// Unified Application Interface
export interface Application {
  id: number;
  first_name: string;
  middle_name?: string;
  last_name: string;
  age?: number;
  sex?: SexEnum;
  date_of_birth: string; // Dates are often handled as strings in TypeScript, assuming ISO format
  license_number?: string;
  status: StatusEnum;
  height?: number;
  
  unit_number? : string;
  street_number? : number;
  street_name? : string;
  po_box? : string;
  city? : string;
  province? : string;
  postal_code? : string;
}

export interface NestedApplication {
  id: number;
  first_name: string;
  middle_name: string;
  last_name: string;
  age: number;
  sex: string;
  date_of_birth: string;
  license_number: string;
  status: string;
  height: number;
  address: {
    unit_number: string | null;
    street_number: number;
    street_name: string;
    po_box: string | null;
    city: string;
    province: string;
    postal_code: string;
  };
}

File: python/app/schemas/application.py
from app.enums.status import StatusEnum
from app.schemas.address import AddressBase
from pydantic import BaseModel
from datetime import date
from typing import Optional
from app.enums import SexEnum, ProvinceEnum
from app.models.address import Address

class ApplicationBase(BaseModel):
    id: Optional[int]
    first_name: str
    middle_name: Optional[str] = None
    last_name: Optional[str] = None
    age: Optional[int] = None
    sex: SexEnum
    height: Optional[int] = None
    date_of_birth: date
    license_number: str
    status: StatusEnum  # Capitalized StatusEnum


class ApplicationCreate(ApplicationBase):
    address: AddressBase  # Nested Address for creation


class ApplicationUpdate(ApplicationBase):
    address: Optional[AddressBase] = None  # Address can be updated

    class Config:
        orm_mode = True


class Application(ApplicationBase):
    is_deleted: bool
    address: AddressBase

    class Config:
        orm_mode = True
